File: LeetCode/101-Symmetric-Tree(Medium)/Solution1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution1 {
    public boolean isSymmetric(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        
        queue.add(root);queue.add(root);
        
        while (!queue.isEmpty()) {
            TreeNode node1 = queue.poll(), node2 = queue.poll();
            
            if (node1 == null && node2 == null) continue;
            if (node1 ==null || node2 == null) return false;
            if (node1.val != node2.val) return false;
            
            queue.add(node1.left);
            queue.add(node2.right);
            
            queue.add(node1.right);
            queue.add(node2.left);
        }
        
        return true;
    }
}

File: LeetCode/116-Populating-Next-Right-Pointers-in-Each-Node(Medium)/Solution1.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() {}

    Node(int _val, Node* _left, Node* _right, Node* _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};

This is not O(1) space
*/
class Solution {
public:
    Node* connect(Node* root) {
        if (root == NULL) return root;

        queue<Node*> q;
        q.push(root);
        
        while(!q.empty()) {
            int size = q.size();
            
            while (size > 0) {
                Node* node = q.front();
                q.pop();
                --size;
                
                if (size == 0) node->next = NULL;
                else node->next = q.front();
                
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
        }
        
        return root;
    }
};

